telethon/telegram_bare_client.py

```diff
@@ -261,7 +272,17 @@ def disconnect(self):
         """
         if new_dc is None:
             # Assume we are disconnected due to some error, so connect again
-            return await self.connect()
+            try:
+                await self._reconnect_lock.acquire()
+                # Another thread may have connected again, so check that first
```

Comment:
There's only one thread with `asyncio` though?

Reply:
Yes and now. In despite of asyncio is one-thread, it's possible that many async calls try to call request (`__call__`) and will `await`ing the result. During this _avaiting_, `_invoke` will call `_reconnect` many times.